File: leetcode/diagonal-traverse.js
// https://leetcode.com/problems/diagonal-traverse/

/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var findDiagonalOrder = function(matrix) {
    var ps = [];
    
    for (var i = 0; i < matrix.length; ++i) {
        for (var j = 0; j < matrix[0].length; ++j) {
            ps.push([i+j, i, matrix[i][j]]);
        }
    }
    
    ps.sort(function (a, b) {
        if (a[0] == b[0]) {
            return (a[1] - b[1]) * (a[0] % 2 == 0 ? -1 : 1);
        } else {
            return a[0] - b[0];
        }
    });
    
    return ps.map(function (p) { return p[2]; });
};

File: leetcode/diagonal-traverse.py
# https://leetcode.com/problems/diagonal-traverse/

class Solution(object):
    def findDiagonalOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if not matrix:
            return []
        
        ans = []
        
        M = len(matrix)
        N = len(matrix[0])
        D = M + N - 1
        
        for d in xrange(D):
            if d % 2 == 1:
                j = min(d, N - 1)
                i = d - j
                s = +1
            else:
                i = min(d, M - 1)
                j = d - i
                s = -1
                
            while 0 <= i < M and 0 <= j < N:
                ans.append(matrix[i][j])
                i += s
                j -= s
        
        return ans
